pkg/tfbridge/provider.go

```diff
@@ -568,88 +568,87 @@ func validateProviderConfig(
-	//			}
-	//		}
-	//	}
-	//} else {
-	//	changes = pulumirpc.DiffResponse_DIFF_NONE
-	//}
-	//
-	//return &pulumirpc.DiffResponse{
-	//	Changes:         changes,
-	//	Replaces:        replaces,
-	//	Diffs:           properties,
-	//	DetailedDiff:    detailedDiff,
-	//	HasDetailedDiff: true,
-	//}, nil
+// A DiffConfig implementation enables the Pulumi provider to detect when provider configuration is
+// changing and suggest a replacement plans when some properties require a replace. Replacing a
+// provider then replaces all resources provisioned against this provider. See also the section on
+// Explicit Provider Configuration in the docs:
+//
+// https://www.pulumi.com/docs/concepts/resources/providers/
+//
+// There is no matching context in Terraform so this remains a Pulumi-level feature.
+func (p *Provider) DiffConfig(
+	ctx context.Context, req *pulumirpc.DiffRequest,
+) (*pulumirpc.DiffResponse, error) {
+	if !p.info.EnableDiffConfig {
+		msg := "DiffConfig is disabled, set ProviderInfo.EnableDiffConfig to enable"
+		return nil, status.Error(codes.Unimplemented, msg)
+	}
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.DiffConfig(%s)", p.label(), urn)
+	glog.V(9).Infof("%s executing", label)
+
+	return plugin.NewProviderServer(&configDiffer{}).DiffConfig(ctx, req)
+}
+
+type configDiffer struct {
+	plugin.UnimplementedProvider
+	schemaMap   shim.SchemaMap
+	schemaInfos map[string]*SchemaInfo
+}
+
+// Schema may specify that changing a property requires a replacement. This honors overrides to
+// ProviderInfo.Schema[x].ForceNew as well as raw schema.ForceNew(), although at the moment
+// realistic TF Providers do not use ForceNew for provider-level properties.
+func (p *configDiffer) isForceNew(path resource.PropertyPath) bool {
+	schemaPath := PropertyPathToSchemaPath(path, p.schemaMap, p.schemaInfos)
+	schema, info, err := LookupSchemas(schemaPath, p.schemaMap, p.schemaInfos)
+	if err != nil {
+		contract.IgnoreError(err)
+		return false
+	}
+	if info != nil && info.ForceNew != nil {
+		return *info.ForceNew
+	}
+	if schema != nil {
+		return schema.ForceNew()
+	}
+	return false
+}
+
+func (p *configDiffer) DiffConfig(
+	urn resource.URN, oldInputs, oldOutputs, newInputs resource.PropertyMap,
+	allowUnknowns bool, ignoreChanges []string,
+) (plugin.DiffResult, error) {
+	contract.Assertf(allowUnknowns, "Expected allowUnknowns to always be true for DiffConfig")
+
+	// Seems that DiffIncludeUnknowns only accepts func (PropertyKey) bool to support ignoring
+	// changes which is awkward for recursive changes, would be better if it supported
+	// func(PropertyPath) bool. Instead of doing this, support IgnoreChanges by copying old
+	// values to new values to disable the diff.
+	newInputsIC, err := propertyvalue.ApplyIgnoreChanges(oldInputs, newInputs, ignoreChanges)
+	if err != nil {
+		return plugin.DiffResult{}, fmt.Errorf("Error applying ignoreChanges: %v", err)
+	}
+
+	objDiff := oldInputs.DiffIncludeUnknowns(newInputsIC)
```

Comment:
Is this appropriate to use here? Does the comment at L624 make sense? It sounds like this function almost supported ignoreChanges but to quite.

Reply:
It looks like a reasonable usage to me.